File: cpp/utils/Image.cpp
#include "Image.h"

#include <fstream>
#include <iostream>

namespace {
Pixel parsePixel(uint32_t inputRgba) {
  Pixel px;
  px.red = static_cast<uint8_t>( (inputRgba & 0xFF000000) >> 24);
  px.green = static_cast<uint8_t>( (inputRgba & 0x00FF0000) >> 16);
  px.blue = static_cast<uint8_t>( (inputRgba & 0x0000FF00) >> 8);
  px.alpha = static_cast<uint8_t>( (inputRgba & 0x000000FF));
  return px;
}
}

PackedImage::PackedImage(const Resolution &res)
    : resolution(res), pixels(res.width * res.height) {

}

bool PackedImage::operator==(const PackedImage &other) const {
  return (resolution == other.resolution) && (pixels == other.pixels);
}

std::istream& operator>>(std::istream &istream, PackedImage &img) {
  uint32_t pixelValue = 0;
  for (Pixel &pixel : img.pixels) {
    istream >> pixelValue;
    pixel = parsePixel(pixelValue);
  }

  return istream;
}

StrideImage::StrideImage(const Resolution &res)
    : resolution(res), redPixels(res.width * res.height),
      greenPixels(res.width * res.height), bluePixels(res.width * res.height),
      alphaPixels(res.width * res.height) {

}

bool StrideImage::operator==(const StrideImage &other) const {
  return (resolution == other.resolution) &&
         (redPixels == other.redPixels) &&
         (greenPixels == other.greenPixels) &&
         (bluePixels == other.bluePixels) &&
         (alphaPixels == other.alphaPixels);
}

std::istream& operator>>(std::istream &istream, StrideImage &img) {
  const int32_t totalPixels = img.resolution.width * img.resolution.height;
  uint32_t pixelValue = 0;
  Pixel currPixel;
  for (int32_t i = 0; i < totalPixels; ++i) {
    istream >> pixelValue;
    currPixel = parsePixel(pixelValue);
    img.redPixels[i] = currPixel.red;
    img.greenPixels[i] = currPixel.red;
    img.bluePixels[i] = currPixel.red;
    img.alphaPixels[i] = currPixel.red;
  }

  return istream;
}

File: python/utils/image.py
#!/usr/bin/env python3

from typing import List
from utils.resolution import Resolution
from utils.pixel import Pixel
from enum import Enum

class ImageType(Enum):
    PackedImageType = 0
    StrideImageType = 1

class PackedImage:
    def __init__(self, resolution: Resolution, pixels: List[Pixel]):
        self.resolution: Resolution = resolution
        self.pixels: List[Pixel] = pixels
    
    def __str__ (self):
        return str(self.resolution) + '\n' + ' '.join(str(p) for p in self.pixels)

    def __eq__(self, other):
        return self.resolution == other.resolution and self.pixels == other.pixels

class StrideImage:
    def __init__(self, resolution: Resolution, pixels: List[Pixel]):
        self.resolution: Resolution = resolution
        self.pixels_red: List[int] = []
        self.pixels_green: List[int] = []
        self.pixels_blue: List[int] = []
        self.pixels_alpha: List[int] = []
        
        self.__split_pixel_components(pixels)

    def merge_pixel_components(self) -> List[Pixel]:
        pixels: List[Pixel] = []
        for i in range(len(self.pixels_red)):
            red: int = self.pixels_red[i]
            green: int = self.pixels_green[i]
            blue: int = self.pixels_blue[i]
            alpha: int = self.pixels_alpha[i]
            pixel: Pixel = Pixel(red, green, blue, alpha)
            pixels.append(pixel)
        return pixels
    
    def __split_pixel_components(self, pixels: List[Pixel]):
        for pixel in pixels:
            self.pixels_red.append(pixel.red)
            self.pixels_green.append(pixel.green)
            self.pixels_blue.append(pixel.blue)
            self.pixels_alpha.append(pixel.alpha)
    
    def __str__ (self):
        return str(self.resolution) + '\n' + \
            ' '.join(str(p) for p in self.pixels_red) + \
            ' '.join(str(p) for p in self.pixels_green) + \
            ' '.join(str(p) for p in self.pixels_blue) + \
            ' '.join(str(p) for p in self.pixels_alpha)

    def __eq__(self, other):
        return self.resolution == other.resolution and \
            self.pixels_red == other.pixels_red and \
            self.pixels_green == other.pixels_green and \
            self.pixels_blue == other.pixels_blue and \
            self.pixels_alpha == other.pixels_alpha

def to_stride_image(image: PackedImage) -> StrideImage:
    return StrideImage(image.resolution, image.pixels)

def to_image(stride_image: StrideImage) -> PackedImage:
    return PackedImage(stride_image.resolution, stride_image.merge_pixel_components())